bears/python/PEP8NotebookBear.py

```diff
@@ -0,0 +1,116 @@
+
+
+def write_notebook_node_to_file(notebook_node):
+    """
+    Writes a NotebookNode to a string array.
+
+    :param notebook_node: The notebook as NotebookNode to write.
+    :return:              The notebook as string array.
+    """
+    notebook_node_as_string_array = (nbformat.writes(notebook_node,
+                                                     nbformat.NO_CONVERT)
+                                     .splitlines(True))
+    return notebook_node_as_string_array
+
+
+class PEP8NotebookBear(LocalBear):
+    LANGUAGES = {"Python", "Python 2", "Python 3"}
+    REQUIREMENTS = {PipRequirement('autopep8', '1.*'),
+                    PipRequirement('nbformat', '4.*')}
+    AUTHORS = {'The coala developers'}
+    AUTHORS_EMAILS = {'coala-devel@googlegroups.com'}
+    LICENSE = 'AGPL-3.0'
+    ASCIINEMA_URL = 'https://asciinema.org/a/83333'
+    CAN_FIX = {'Formatting'}
+
+    def run(self, filename, file,
+            max_line_length: int=79,
+            indent_size: int=SpacingHelper.DEFAULT_TAB_WIDTH,
+            pep_ignore: typed_list(str)=(),
+            pep_select: typed_list(str)=(),
+            local_pep8_config: bool=False):
+        """
+        Detects and fixes PEP8 incompliant code in Jupyter Notebooks. This bear
+        will not change functionality of the code in any way.
+
+        :param max_line_length:   Maximum number of characters for a line.
+        :param indent_size:       Number of spaces per indent level.
+        :param pep_ignore:        A list of errors/warnings to ignore.
+        :param pep_select:        A list of errors/warnings to exclusively
+                                  apply.
+        :param local_pep8_config: Set to true if autopep8 should use a config
+                                  file as if run normally from this directory.
+        """
+        options = {"ignore": pep_ignore,
+                   "select": pep_select,
+                   "max_line_length": max_line_length,
+                   "indent_size": indent_size}
+        notebook_node = read_notebook_node_from_file(file)
+        cells = notebook_node['cells']
+
+        def autopep8_fix_code(source):
+            source_corrected = \
+                autopep8.fix_code(source,
+                                  apply_config=local_pep8_config,
+                                  options=options)
+            # autopep8.fix_code automatically adds a final newline at the end,
+            # e.g. `autopep8.fix_code('a=1')` yields 'a = 1\n'.
+            # This is not related to the 'W292' flag, i.e.
+            # `autopep8.fix_code('a=1', options=dict(ignore=('W292',)))` gives
+            # the same result.
+            # For notebook cells, this behaviour does not make sense, hence
+            # newline is removed if `source` does not end with one.
+            if not source.endswith('\n'):
+                return source_corrected[:-1]
+            return source_corrected
```

Comment:
hm might it be cleaner to have this outside the class and pass the variables explicitly rather than in the closure? https://www.python.org/dev/peps/pep-0020/